[microsoft-ui-uiautomation] Add new port

srktirumala opened the pull request:
<!-- If your PR fixes issues, please note that here by adding "Fixes #NNNNNN." for each fixed issue on separate lines. -->

<!-- If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/. -->

<!-- If this PR updates an existing port, please uncomment and fill out this checklist:

- [ ] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [ ] SHA512s are updated for each updated download.
- [ ] The "supports" clause reflects platforms that may be fixed by this new version.
- [ ] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [ ] Any patches that are no longer applied are deleted from the port's directory.
- [ ] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [ ] Only one version is added to each modified port's versions file.

END OF PORT UPDATE CHECKLIST (delete this line) -->

If this PR adds a new port, please uncomment and fill out this checklist:

- [ x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md). 
- [ x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines. 
- [ x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [ x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [ x] The license declaration in `vcpkg.json` matches what upstream says.
- [ x] The installed as the "copyright" file matches what upstream says.
- [ x] The source code of the component installed comes from an authoritative source.
- [ x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [ x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [ x] Only one version is in the new port's versions file.
- [ x] Only one version is added to each modified port's versions file.

shtirumaMSFT committed:
new vcpkg for UIA

srktirumala committed:
fixes to manifest file vcpkg.json for parse errors

srktirumala committed:
fixing sha in json files

srktirumala committed:
some fixes

srktirumala committed:
some port fixes

srktirumala committed:
fixes

srktirumala committed:
fixing dynamic link issue on port file

srktirumala commented:
@jimwang118 ,
 I need help with fixes.
The port for which I am writing this is only available for Windows. so I do not want this to run for other platforms. how do I fix it in my portfile so that other builds like for android etc. do not run..
as of now I see that x64 Windows build is green.. and I think that is enough for me to consume from vcpkg at this point in time.. is that possible to make other platforms skip build for this specific port?

srktirumala committed:
fixes to versions

srktirumala committed:
Update version database

srktirumala committed:
Update vcpkg json

srktirumala committed:
fixes to sha

srktirumala committed:
updated sha for version file

srktirumala committed:
vrsion file fixes

srktirumala committed:
git-tree fix

srktirumala committed:
win32 build config

srktirumala committed:
git tree fix

srktirumala committed:
fix build platforms

srktirumala committed:
fix git tree

srktirumala committed:
fixing platform

srktirumala committed:
git tree fix

srktirumala committed:
removing platform

srktirumala committed:
git tree fix

srktirumala committed:
platform

srktirumala committed:
git tree fix

srktirumala committed:
config fixes

srktirumala committed:
git tree fix

jimwang118 commented on versions/m-/microsoft-ui-uiautomation.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "versions": [
+    {
+      "version-date": "2022-04-26",
+      "git-tree": "d708bf06334c1ff88eb0b7f671e845ac3efb3af0"
+    }
+  ]
+}
```
No manual writing is required. Executing the command `./vcpkg x-add-version {port}` will automatically generate the file.

srktirumala committed:
changing the build project as needed for us to vcpkg

srktirumala committed:
git tree has fix

srktirumala committed:
changing the build project as needed for us to vcpkg

srktirumala committed:
git tree has fix

jimwang118 commented on versions/baseline.json:
```diff
@@ -5672,6 +5672,10 @@
       "baseline": "0.1.0-alpha4",
       "port-version": 8
     },
+    "microsoft-ui-uiautomation": {
+      "baseline": "2022-04-26",
+      "port-version": 0
+    },
```
There is no need to modify it manually, it will be automatically generated using the command ./vcpkg x-add-version {port}. If you don't understand anything about the command, you can check the [documentation](https://github.com/microsoft/vcpkg-docs).

jimwang118 commented on ports/microsoft-ui-uiautomation/vcpkg.json:
```diff
@@ -0,0 +1,14 @@
+  "version-date": "2022-04-26",
+  "description": "Microsoft UI Automation library",
+  "homepage": "https://github.com/microsoft/Microsoft-UI-UIAutomation",
+  "license": "MIT",
```
```suggestion
  "license": "MIT",
   "supports": "windows",
```

srktirumala committed:
Update ports/microsoft-ui-uiautomation/vcpkg.json

Co-authored-by: jim wang <122244446+jimwang118@users.noreply.github.com>

srktirumala committed:
git tree fix

srktirumala committed:
format json

srktirumala commented:
@MonicaLiu0311, I see this error during local build and install:

 1>Project "D:\vcpkgUIA\buildtrees\microsoft-ui-uiautomation\x64-windows-rel\UiaOperationAbstraction\UiaOperationAbstraction.vcxproj" on node 1 (Restore target(s)).
 1>_GetAllRestoreProjectPathItems:
     Determining projects to restore...
   Restore:
     X.509 certificate chain validation will use the default trust store selected by .NET for code signing.
     X.509 certificate chain validation will use the default trust store selected by .NET for timestamping.
     Restoring NuGet package Microsoft.Windows.CppWinRT.2.0.200703.9.
     Restoring NuGet package Microsoft.Windows.ImplementationLibrary.1.0.191107.2.
     Adding package 'Microsoft.Windows.ImplementationLibrary.1.0.191107.2' to folder 'D:\vcpkgUIA\buildtrees\microsoft-ui-uiautomation\x64-windows-rel\UiaOperationAbstraction\packages'
     Adding package 'Microsoft.Windows.CppWinRT.2.0.200703.9' to folder 'D:\vcpkgUIA\buildtrees\microsoft-ui-uiautomation\x64-windows-rel\UiaOperationAbstraction\packages'
     Added package 'Microsoft.Windows.CppWinRT.2.0.200703.9' to folder 'D:\vcpkgUIA\buildtrees\microsoft-ui-uiautomation\x64-windows-rel\UiaOperationAbstraction\packages'
     Added package 'Microsoft.Windows.ImplementationLibrary.1.0.191107.2' to folder 'D:\vcpkgUIA\buildtrees\microsoft-ui-uiautomation\x64-windows-rel\UiaOperationAbstraction\packages'
     
     NuGet Config files used:
         C:\Users\shtiruma\AppData\Roaming\NuGet\NuGet.Config
         C:\Program Files (x86)\NuGet\Config\Microsoft.VisualStudio.FallbackLocation.config
         C:\Program Files (x86)\NuGet\Config\Microsoft.VisualStudio.Offline.config
     
     Feeds used:
         C:\Users\shtiruma\.nuget\packages\
         https://api.nuget.org/v3/index.json
         C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\
     
     Installed:
         2 package(s) to packages.config projects
 1>Done Building Project "D:\vcpkgUIA\buildtrees\microsoft-ui-uiautomation\x64-windows-rel\UiaOperationAbstraction\UiaOperationAbstraction.vcxproj" (Restore target(s)).

I am not sure why the vcpkg msbuild is using a different working directory, other than the UiaOperationAbstraction/UiaOperationAbstraction.vcxproj available under ${SOURCE_PATH}/src/UIAutomation/.. where the required nuget packages are getting downloaded to.

srktirumala committed:
code chnages

srktirumala committed:
git tree fix

srktirumala commented:
@MonicaLiu0311 , could you please help me address the issue i posted above.

jimwang118 commented:
1. First of all, it is wrong to obtain the relative path of winrt package in your project `UiaOperationAbstraction.vcxproj`. The relative path of winrt corresponding to this project should be `packages\Microsoft.Windows.CppWinRT.2.0.200703.9\build\native\Microsoft.Windows.CppWinRT.props`.
2. Does compiling the UiaOperationAbstraction.vcxproj project depend on other project codes in UIAutomation.sln?

srktirumala commented:
@jimwang118 , UiaOperationAbstraction project binaries is what we consume as part of OSS in Windows OS project. The solution contains 3 project out of which 1 is the test project [SharpConsoleAppDemo]. I donot want to package all the other binaries on the solution for vcpkg. hence trying to build only UiaOperationAbstraction in the portfile.
Is that correct understanding?

jimwang118 commented:
I installed `nuget targets and build tasks` in local VS.
Then delete the `test` and `demo` projects in` UIAutomation.sln` and compile directly. The compilation result is normal.

If you compile UiaOperationAbstraction.vcxproj directly and fix the winrt-dependent path in the project as follows, the compilation will prompt `Cannot open include file: 'winrt/Microsoft.UI.UIAutomation.h'`.
![image](https://github.com/microsoft/vcpkg/assets/122244446/cbeebd40-22c2-4c96-8c71-1a56a4cec4bc)

srktirumala commented:
@jimwang118 , let me try updating the solution file and try. to close it faster.

srktirumala commented:
@jimwang118 , how do i refer the solution file or vcxproj file which i modified locally on teh portfile.cmake to compile vcpkg with local changes ?

jimwang118 commented:
> @jimwang118 , how do i refer the solution file or vcxproj file which i modified locally on teh portfile.cmake to compile vcpkg with local changes ?

The main thing you have to consider is how to download `Microsoft.Windows.CppWinRT.2.0.200703.9` and `Microsoft.Windows.ImplementationLibrary.1.0.191107.2` and copy them to the correct directory `F:\vcpkgUIA\buildtrees\microsoft-ui-uiautomation\x64-windows-rel\packages`.
If other projects can pass the local test, they can be used directly without modification.

jimwang118 commented:
@srktirumala Thanks for your PR, Is work still being done for this PR?

jimwang118 commented:
Closing this PR since it seems that no progress is being made. Please ping us to reopen if work is still being done.
